File: typescripts/nonce_account.ts
import {
    Connection,
    PublicKey,
    Keypair,
    NonceAccount,
    Transaction,
    sendAndConfirmTransaction,
    TransactionInstruction,
    SystemProgram
} from '@solana/web3.js';
import * as fs from 'fs';
import * as dotenv from 'dotenv';
dotenv.config();
function readKeypair(file: string): Keypair {
    const keypairPath = `${process.env.HOME}/.config/solana/${file}`;
    const keypairData = JSON.parse(fs.readFileSync(keypairPath, 'utf8'));
    return Keypair.fromSecretKey(new Uint8Array(keypairData));
}
(async () => {
    const connection = new Connection('https://api.devnet.solana.com', 'confirmed');
    const payer = readKeypair("id.json");
    const nonceAccountKey = readKeypair("nonce-account.json")
    const accountInfo = await connection.getAccountInfo(nonceAccountKey.publicKey);
    if (accountInfo === null) throw new Error("")
    const nonceAccount = NonceAccount.fromAccountData(accountInfo.data);
    const programId = new PublicKey(process.env.program_id as string);

    // make a nonce advance instruction
    const advanceIX = SystemProgram.nonceAdvance({
        authorizedPubkey: payer.publicKey,
        noncePubkey: nonceAccountKey.publicKey,
    });
    const ix = new TransactionInstruction({
        // our program only logs hello world and doesn't need any accounts.
        keys: [{pubkey: payer.publicKey, isSigner: true, isWritable: true}],
        programId,
        data: Buffer.alloc(0)
    });
    const tx = new Transaction();
    tx.add(advanceIX);
    tx.add(ix);
    tx.recentBlockhash = nonceAccount.nonce;
    tx.feePayer = payer.publicKey;

    const signature = await sendAndConfirmTransaction(
        connection,
        tx,
        [payer]
    );
    console.log('signature/txhash:', signature);
})();

File: examples/nonce_account.rs
use solana_client::rpc_client::RpcClient;
use solana_sdk::{
    commitment_config::CommitmentConfig,
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
    signature::Signer,
    transaction::Transaction,
};
fn main() {
    dotenv::dotenv().unwrap();
    let client = RpcClient::new_with_commitment(
        "https://api.devnet.solana.com".to_string(),
        CommitmentConfig::confirmed(),
    );
    let payer = solana_client_example::keypair();
    let nonce_account = solana_client_example::nonce_account();
    let nonce_account_data = client.get_account(&nonce_account.pubkey()).unwrap();
    let nonce_data = solana_rpc_client_nonce_utils::data_from_account(&nonce_account_data).unwrap();
    let block_hash = nonce_data.blockhash();

    let program_id: Pubkey = std::env::var("hello_world").unwrap().parse().unwrap();

    let advance_ix = solana_sdk::system_instruction::advance_nonce_account(
        &nonce_account.pubkey(),
        &payer.pubkey(),
    );
    let instruction = Instruction {
        program_id,
        accounts: vec![AccountMeta::new(payer.pubkey(), true)],
        data: Vec::new(),
    };
    let transaction = Transaction::new_signed_with_payer(
        &[advance_ix, instruction],
        Some(&payer.pubkey()),
        &[&payer],
        block_hash,
    );
    match client.send_and_confirm_transaction(&transaction) {
        Ok(signature) => println!("Transaction sent with signature: {}", signature),
        Err(err) => eprintln!("Error sending transaction: {:?}", err),
    }
}
